### PR TITLE
Add join, leave and kill subnet method

### PR DESCRIPTION
**Pull request by cryptoAtwill**
Add `join`, `leave` and `kill` subnet json rpc handlers.

### CODE REVIEW

File: src/server/handlers/join.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+// Copyright 2022-2023 Protocol Labs
```
Comment (adlrocha) [inline src/server/handlers/join.rs:1]: I think we should group handles by topical modules. In this case, add, join, kill and leave should be part of a subnet or manager module, WDYT?

File: src/server/handlers/join.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    }
+}
+
+#[async_trait]
+impl JsonRPCRequestHandler for JoinSubnetHandler {
+    type Request = JoinSubnetParams;
+    type Response = ();
+
+    async fn handle(&self, request: Self::Request) -> anyhow::Result<Self::Response> {
+        let conn = match self.pool.get(&request.subnet) {
```
Comment (adlrocha) [inline src/server/handlers/join.rs:43]: This is a bug, you are getting the connection to the subnet ID, not to the parent, and you want to interact with the subnet actor in the parent.

File: src/server/handlers/kill.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    }
+}
+
+#[async_trait]
+impl JsonRPCRequestHandler for KillSubnetHandler {
+    type Request = KillSubnetParams;
+    type Response = ();
+
+    async fn handle(&self, request: Self::Request) -> anyhow::Result<Self::Response> {
+        let conn = match self.pool.get(&request.subnet) {
```
Comment (adlrocha) [inline src/server/handlers/kill.rs:39]: Same as above, you should get the conn for the parent of the subnet here.

File: src/server/handlers/leave.rs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    }
+}
+
+#[async_trait]
+impl JsonRPCRequestHandler for LeaveSubnetHandler {
+    type Request = LeaveSubnetParams;
+    type Response = ();
+
+    async fn handle(&self, request: Self::Request) -> anyhow::Result<Self::Response> {
+        let conn = match self.pool.get(&request.subnet) {
```
Comment (adlrocha) [inline src/server/handlers/leave.rs:39]: Same as above, you should get the conn for the parent of the subnet here.

File: src/server/handlers/mod.rs [hunk 5]
```diff
@@ -53,29 +100,25 @@ impl Handlers {
...
         Ok(Self { handlers })
     }
-
-    pub async fn handle(&self, method: Method, params: Value) -> Result<Value> {
-        if let Some(wrapper) = self.handlers.get(&method) {
-            match wrapper {
-                HandlerWrapper::CreateSubnet(handler) => {
-                    let r = handler.handle(serde_json::from_value(params)?).await?;
-                    Ok(serde_json::to_value(r)?)
-                }
```
Comment (adlrocha) [inline src/server/handlers/mod.rs:71]: I am wondering if the body for each of this branches can be de-duplicated without the need of a macro, it is the same for all of them

File: src/server/handlers/mod.rs [hunk 6]
```diff
@@ -18,19 +24,60 @@ use serde_json::Value;
 use std::collections::HashMap;
 use std::sync::Arc;
 
-pub type Method = String;
+/// A util macro to create the handlers and do the method routing.
+/// This macro generates the `handle` method. In the method, it just matches the handler with
+/// different enum types to get the inner handler instance. Every time a new handler is added,
+/// a new match statement must be copy-pasted. With this macro, we can reduce a lot of repetitive
+/// copy paste.
+macro_rules! route_handlers {
```
Comment (adlrocha) [inline src/server/handlers/mod.rs:32]: Would you mind keeping this PR only for the implementation of the handlers and move the macro proposal to another PR? It may potentially require a longer discussion and I don't want to block these commands from being merged. Thanks!

**Review by adlrocha — CHANGES_REQUESTED** [review]
Amazing work. A few bugs, and a proposal to move the macro to another PR to avoid delaying these commands due to a discussion over the macro. Thanks.

File: src/server/handlers/mod.rs (hunk 6 above)
Comment (cryptoAtwill) [reply src/server/handlers/mod.rs:32]: sure sure

File: src/server/handlers/leave.rs (hunk 4 above)
Comment (cryptoAtwill) [reply src/server/handlers/leave.rs:39]: updated!

File: src/server/handlers/kill.rs (hunk 3 above)
Comment (cryptoAtwill) [reply src/server/handlers/kill.rs:39]: updated!

File: src/server/handlers/join.rs (hunk 2 above)
Comment (cryptoAtwill) [reply src/server/handlers/join.rs:43]: updated!

File: src/server/handlers/join.rs (hunk 1 above)
Comment (cryptoAtwill) [reply src/server/handlers/join.rs:1]: updated and placed in a separate mod.

File: src/server/handlers/mod.rs (hunk 5 above)
Comment (cryptoAtwill) [reply src/server/handlers/mod.rs:71]: No need enum and macro now, I found a way to handle it already!

File: src/server/handlers/join.rs [hunk 7]
```diff
@@ -40,15 +40,19 @@ impl JsonRPCRequestHandler for JoinSubnetHandler {
     type Response = ();
 
     async fn handle(&self, request: Self::Request) -> anyhow::Result<Self::Response> {
-        let conn = match self.pool.get(&request.subnet) {
+        let subnet = SubnetID::from_str(&request.subnet)?;
+        let parent = subnet
+            .parent()
+            .ok_or_else(|| anyhow!("no parent found"))?
+            .to_string();
+        let conn = match self.pool.get(&parent) {
```
Comment (adlrocha) [inline src/server/handlers/join.rs]: Wouldn't it make more sense for `pool.get` to accept a `SubnetID` as an argument instead of a string? It would remove the need of an additional check inside of get and having to do a lot of back and forth between types.

**Review by adlrocha — APPROVED** [review]
LGTM. Just one last comment regarding using `SubnetID` as the argument type to `pool.get`. Thanks.

File: src/server/handlers/join.rs (hunk 7 above)
Comment (cryptoAtwill) [reply src/server/handlers/join.rs]: yeah, make sense, haha. Update this now.
